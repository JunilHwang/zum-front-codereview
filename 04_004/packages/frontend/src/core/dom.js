import { instantiateTree } from './reconciler';

export class Component {
  constructor(props = {}) {
    this.props = props;
    this.state = {};
  }

  setState(partialState) {
    this.state = {
      ...this.state,
      ...partialState,
    };
  }
}
Component.prototype.isComponent = {};

/* istanbul ignore next */
function warnIfKeyNotExists(children) {
  if (
    children.some(child => typeof child.props.key === 'undefined') ||
    new Set(children.map(child => child.props.key)).size !== children.length
  ) {
    console.warn('children of list should have unique keys');
  }
}

function childToString(child) {
  return ['object', 'function'].includes(typeof child) ?
    child :
    String(child);
}

function wrapChildren(children) {
  /* istanbul ignore next */
  if (__DEV__ && children.some(child => Array.isArray(child))) {
    children.forEach(warnIfKeyNotExists);
  }
  return children.flat().map(childToString);
}

export function createElement(type, props, ...children) {
  return {
    type,
    props: Object
      .assign(
        {},
        props,
        children.length > 0 &&
        { children: wrapChildren(children) } ||
        { children }),
  };
}

export function createRef() {
  return {current: null};
}

function clearContainer(container) {
  container.childNodes
    .forEach(container.removeChild.bind(container));
}

export function render(element, container) {
  const instance = instantiateTree(element);
  if (container._mountedInstanceTree) {
    const mounted = container._mountedInstanceTree;
    mounted.diff(element);
    return;
  }
  clearContainer(container);

  const rootNode = instance.mount();
  container._mountedInstanceTree = instance;
  container.appendChild(rootNode);
}
