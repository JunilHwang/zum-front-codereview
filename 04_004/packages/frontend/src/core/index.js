export * from './context';
export * from './dom';
export * from './router';
export * from './hoc';
export * from './util';
