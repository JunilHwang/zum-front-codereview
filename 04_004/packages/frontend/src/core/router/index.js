export * from './match';
export * from './request';
export * from './wrapper';
export * from './components';
