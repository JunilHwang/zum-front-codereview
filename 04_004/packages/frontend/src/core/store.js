class Store {
  constructor(reducer) {
    this.reducer = reducer;
    this.state = reducer();
    this.subscriber = new WeakSet();
  }

  getState() {
    return this.state;
  }

  subscribe(callback) {
    this.subscriber.add(callback);
    return () => this.subscriber.delete(callback);
  }

  dispatch(action) {
    this.state = this.reducer(this.state, action);
    this.subscriber.forEach(callback => callback(action));
  }
}

export function createStore(reducer) {
  return new Store(reducer);
}
