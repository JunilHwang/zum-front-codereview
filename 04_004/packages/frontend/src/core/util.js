export class Cache {
  set(key, value) {
    const item = {
      type: typeof value,
      value,
    };
    return localStorage.setItem(key, JSON.stringify(item));
  }

  get(key) {
    if (Cache.prototype.has.call(this, key)) {
      const cache = localStorage.getItem(key);
      try {
        const item = JSON.parse(cache);
        if (item.type === typeof item.value) {
          return item.value;
        }
      } catch (e) {}
    }
    return null
  }

  has(key) {
    return localStorage.getItem(key) !== null;
  }

  clear() {
    return localStorage.clear();
  }

  delete(key) {
    return localStorage.removeItem(key);
  }
}

export function formatTimestamp(timestamp) {
  const date = new Date(timestamp);
  return [date.getFullYear(), date.getMonth()+1, date.getDate()].join('/');
}

export function compose(...funcs) {
  return function (...args) {
    const init = funcs.pop()(...args);
    return funcs
      .reverse()
      .reduce((acc, f) => f(acc), init);
  };
}
