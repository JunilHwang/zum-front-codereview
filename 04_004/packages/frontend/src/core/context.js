import { Component } from './dom';

const store = new WeakMap();
export function createContext(defaultValue) {
  class Provider extends Component {
    render() {
      store.set(Provider, this.props.value);
      return this.props.children[0];
    }
  }

  class Consumer extends Component {
    render() {
      const value = store.has(Provider) ? store.get(Provider) : defaultValue;
      return this.props.children[0](value);
    }
  }

  return {
    Provider,
    Consumer,
  };
}
