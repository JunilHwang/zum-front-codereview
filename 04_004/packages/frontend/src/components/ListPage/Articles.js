import { createElement as h, Component, Link, formatTimestamp } from '@/core';

export class Articles extends Component {
  render() {
    /** @type {Article[]} */
    const articles = this.props.articles;
    return (
      h('table', null,
        h('thead', null,
          h('tr', null,
            h('th', null, '글번호'),
            h('th', null, '제목'),
            h('th', null, '작성자'),
            h('th', null,
              h('a', {href: '#', onClick: this.props.resort.bind(this)}, '작성일')
            ),
          )
        ),
        h('tbody', null,
          articles.map(article => (
            h('tr', {key: article.id},
              h('th', null, String(article.id)),
              h('td', null,
                h(Link, {to: `/${article.id}`}, article.title)
              ),
              h('td', null,
                h('a', {
                  href: '#',
                  onClick: (event) => {
                    event.preventDefault();
                    this.props.search({
                      searchField: 'author',
                      keyword: article.author,
                      page: 0,
                    });
                  }
                }, article.author)
              ),
              h('td', null, formatTimestamp(article.timestamp)),
            )
          ))
        )
      )
    );
  }
}
