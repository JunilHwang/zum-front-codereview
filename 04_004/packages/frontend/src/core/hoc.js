import { createElement as h, Component } from './dom';
import { request } from '@/core/router';

export function withFetch(url, options = {}) {
  return function (WrappedComponent) {
    return class extends Component {
      constructor(props) {
        super(props);

        this.state = {
          error: null,
          data: null,
        };
        this.abortController = null;
      }

      componentWillUnmount() {
        try {
          this.abortController?.abort();
        } catch (e) {
          console.error(e);
        }
      }

      fetchData() {
        if (this.abortController) return;
        this.abortController = new AbortController();

        request(
          typeof url === 'function' ?
          url(this.props) :
          url, {
            ...options,
            signal: this.abortController.signal
          })
          .then(result => this.setState({ data: result }))
          .catch(err => {
            if (500 <= err.status) {
              // TODO: alert 컴포넌트 만들기
              alert(`오류(${err.status}): ${err.result.message}`);
            }
            this.setState({ error: err, data: err.result });
          })
          .finally(() => this.abortController = null);
      }

      render() {
        return h(WrappedComponent, {
          fetchedData: this.state.data,
          error: this.state.error,
          fetch: this.fetchData.bind(this),
          abort: this.abortController?.abort.bind(this.abortController) ||
            (() => {}),
          ...this.props
        });
      }
    };
  };
}

export function withInitFetch(url, options = {}) {
  return function (WrappedComponent) {
    return withFetch(url, options)(
      class extends Component {
        componentDidMount() {
          this.props.fetch();
        }

        componentWillUnmount() {
          this.props.abort();
        }

        render() {
          return h(WrappedComponent, {...this.props});
        }
      }
    );
  }
}
