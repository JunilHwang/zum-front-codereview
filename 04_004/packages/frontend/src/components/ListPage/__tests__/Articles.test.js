/**
 * @jest-environment jsdom
 */
import { fireEvent, getByText } from '@testing-library/dom';
import { createElement as h, render } from '@/core';
import { Articles } from '../Articles';

let container;
beforeEach(() => {
  container = document.createElement('div');
  jest.useFakeTimers();
});

afterEach(() => {
  container.innerHTML = '';
  jest.useRealTimers();
});

const act = callback => {
  callback();
  jest.runAllTimers();
};

const articles = [
  {id: 1, title: 'foo', content: 'bar', author: 'baz', timestamp: 0}
];
test('게시글 목록 렌더링', () => {
  render(h(Articles, {articles, resort: () => {}}), container);
  expect(container).toMatchSnapshot();
});

test('작성일을 눌러 정렬 수행', () => {
  const mock = jest.fn();
  render(h(Articles, {articles, resort: mock}), container);
  act(() => {
    fireEvent.click(getByText(container, '작성일', {selector: 'a'}));
  });
  expect(mock).toBeCalled();
});

test('작성자를 눌러 검색', () => {
  const mock = jest.fn();
  render(
    h(Articles, {
      articles,
      resort: () => {},
      search: mock
    }),
    container
  );
  act(() => {
    fireEvent.click(getByText(container, 'baz', {selector: 'a'}));
  });
  expect(mock).toBeCalledWith({keyword: 'baz', page: 0, searchField: 'author'});
});