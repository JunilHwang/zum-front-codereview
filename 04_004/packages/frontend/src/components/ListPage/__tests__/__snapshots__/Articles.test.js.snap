// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`게시글 목록 렌더링 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th>
          글번호
        </th>
        <th>
          제목
        </th>
        <th>
          작성자
        </th>
        <th>
          <a
            href="#"
          >
            작성일
          </a>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>
          1
        </th>
        <td>
          <a
            href="/1"
          >
            foo
          </a>
        </td>
        <td>
          <a
            href="#"
          >
            baz
          </a>
        </td>
        <td>
          1970/1/1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
